server/api/impl/admin/synchronize.go

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package admin
+
+import (
+	"encoding/json"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/mattermost/mattermost-plugin-apps/server/api"
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/pkg/errors"
+)
+
+// appsMappingsEnvVarName determines an environment variable.
+// Variable saves address of installation's app mapping's file.
+// File is given by a bucket name and key delimited by a `/`
+// example: export MM_APPS_MAPPINGS = "apps/latest_mappings"
+const appsMappingsEnvVarName = "MM_APPS_MAPPINGS"
+
+const oldVersionKey = "update_from_version"
+
+// Mappings describes mappings for all the apps in this installation
+// for now it supports lambda functions and static assets
+// appID -> app manifest
+type Mappings struct {
+	Apps map[api.AppID]*api.Manifest
+}
+
+// MappingsFromJSON deserializes a list of app mappings from json.
+func MappingsFromJSON(data []byte) (Mappings, error) {
+	var mappings Mappings
+	if err := json.Unmarshal(data, &mappings); err != nil {
+		return Mappings{}, err
+	}
+	return mappings, nil
+}
+
+// ToJSON serializes a list of app mappings to json.
+func (m *Mappings) ToJSON() []byte {
+	b, _ := json.Marshal(m)
+	return b
+}
+
+// SynchronizeApps synchronizes apps upgrading, downgrading and deleting apps
+func (adm *Admin) SynchronizeApps() error {
```

Comment:
2/5 I still maintain that tracking the previous version, and providing upgrade/downgrade messages from the proxy plugin is premature.

If we do want to track the state and detect these state transitions, we need to deduplicate the App messages in HA, avoid sending 1 message/server.

Reply:
It's just the comment left from before. upgrade/downgrade messages are not sent. Only `onStartup` and `onDelete` - so there is no duplication.